Adding onboarding docs and configuring live preview (#5333)
ROCm/rocm-libraries: projects/hipdnn/docs/user-guides/how-to/migrate-cudnn.rst

.. meta::
  :description: Learn how to migrate your frontend code from cuDNN to hipDNN.
  :keywords: hipDNN, ROCm, migrate, cuDNN

.. _migrate-cudnn:

*********************************

> Migrate a cuDNN project to hipDNN

Reviewer: ```suggestion
Migrate NVIDIA CUDA cuDNN projects to hipDNN
```

Reviewer: first reference in title need to provide proper third-party attribution/naming/trademarking

Author: added

*********************************

> This guide demonstrates how to migrate a cuDNN code project to hipDNN.

Reviewer: ```suggestion
This topic demonstrates how to migrate a NVIDIA CUDA cuDNN project to hipDNN.
```

Reviewer: From here on just use cuDNN

> Before you begin, ensure hipDNN (ROCm) is installed. See :ref:`prerequisites` for more information.

Reviewer: ```suggestion
Before you begin, ensure hipDNN is installed. See :ref:`prerequisites` for more information.
```

Reviewer: Is this sentence even necessary?

Here's a minimal example of a hipDNN project in ``CMakeLists.txt``:

.. code:: cmake

    cmake_minimum_required(VERSION <your_minimum>)
    project(my_app LANGUAGES CXX)
    set(CMAKE_CXX_STANDARD 17)

    find_package(hipdnn_frontend CONFIG REQUIRED)

    add_executable(my_app main.cpp)
    target_link_libraries(my_app PRIVATE hipdnn_frontend)

.. tip::

  See `Working examples in the Porting Guide <https://github.com/ROCm/rocm-libraries/blob/develop/projects/hipdnn/docs/PortingGuide.md#working-examples>`_ for ported code samples.

> Key differences between cuDNN and hipDNN

Reviewer: Are we sure want cu first and hip second in our comparisons and text?

Author: I think it's consistent with the rest of the article. We're going from cu to hip.

========================================

> Review this table to see a high-level overview of the differences between cuDNN and hipDNN:

Reviewer: ```suggestion
This table provides a high-level overview of the differences between cuDNN and hipDNN:
```

.. list-table::
   :widths: 3 3 3
   :header-rows: 1

>    * - Aspect

Reviewer: Is this the right term

Author: I believe it is but @CMiservaAMD  can you confirm?

Reviewer 2: "Aspect" seems most suitable to me, yes.

   * - Handle destruction
     - ``cudnnDestroy(handle)``
     - ``hipdnnDestroy(handle)``
   * - Heuristics modes
     - All cuDNN heuristic modes
     - Currently only ``HeurMode_t::FALLBACK``
   * - Operation support
     - All cuDNN operations
     - See :ref:`plugin-support` for more information.
   * - Device memory utility
     - ``Surface<type>``
     - ``MigratableMemory<type>``
   * - Device memory access
     - ``Surface<type>::devPtr``
     - ``MigratableMemory<type>::deviceData()``

Tensor dimensions and layouts
=============================

Tensor dimension ordering in hipDNN is *operation-specific*, following the same conventions as
PyTorch and cuDNN. The *memory layout* (channel-first vs channel-last) is always controlled by
strides and stride order, not by the order of the tensor dimension vector that always holds values as [N,C,H,W] or [N,C,D,H,W]. 

For example, memory arranged as NCHW corresponds to stride order {3,2,1,0} (W is the most tightly packed), and NDHWC corresponds to stride order {4,0,3,2,1} (C is the most tightly packed). 
Use the ``TensorLayout`` constants and the ``generateStrides()`` utility to compute strides for common layouts.

> 

Reviewer: It's not clear to me what the purpose of these tables are?

<img width="406" height="312" alt="Image" src="https://github.com/user-attachments/assets/de7abc84-0e27-4978-bf41-0874fa753de4" />

Is it to show the differences with cuDNN?

Author: @CMiservaAMD  do you think we can add some contextual information before these headings to make this clearer?

Reviewer 2: Reviewing this content further, I think it would be more appropriate to have this in the Operation Support page. I relocated this content to that page in https://github.com/ROCm/rocm-libraries/pull/5638

   :widths: 3 3 5
   :header-rows: 1

   * - Tensor
     - Shape
     - Description
   * - A
     - ``(...batch, M, K)``
     - Leading batch dims, last two are matrix dims
   * - B
     - ``(...batch, K, N)``
     - K must match A's last dim
   * - C (output)
     - ``(...batch, M, N)``
     - Batch dims are broadcast

Batch dimensions support broadcasting (dims must be equal or divisible).

.. code:: cpp

  // Matmul example: A(batch, M, K) @ B(batch, K, N) -> C(batch, M, N)
  auto a = TensorAttributes()
              .set_dim({4, 128, 64})    // batch=4, M=128, K=64
              .set_stride({128*64, 64, 1});

  auto b = TensorAttributes()
              .set_dim({4, 64, 256})    // batch=4, K=64, N=256
              .set_stride({64*256, 256, 1});

Pointwise operations
--------------------

Pointwise operations (ReLU, Sigmoid, Add, Mul, and so on) are *dimension-agnostic* — they accept
tensors of any shape. For binary and ternary operations, inputs are broadcast using NumPy-style
broadcasting rules (dimensions compared right-to-left; compatible if equal or 1).

Troubleshooting
===============

Error: Missing Heuristic modes A and B
--------------------------------------

> The heuristic implementation in hipDNN has yet to be implemented

Reviewer: ```suggestion
The heuristic implementation in hipDNN is not implemented.
```

Reviewer: seems like a promise it would be
